src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
 namespace kaleidoscope {
 namespace plugin {
 
-// ---- state ---------
+// ----------------------------------------------------------------------------
+// Configuration variables
 
-uint16_t OneShot::start_time_ = 0;
 uint16_t OneShot::time_out = 2500;
 uint16_t OneShot::hold_time_out = 250;
 int16_t OneShot::double_tap_time_out = -1;
-OneShot::key_state_t OneShot::state_[OneShot::ONESHOT_KEY_COUNT];
-Key OneShot::prev_key_;
-bool OneShot::should_cancel_ = false;
-bool OneShot::should_cancel_stickies_ = false;
 
-bool OneShot::isPressed() {
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT; i++) {
-    if (state_[i].pressed)
+// ----------------------------------------------------------------------------
+// State variables
+
+OneShot::OneShotKeyState OneShot::state_[OneShot::ONESHOT_KEY_COUNT];
+uint16_t OneShot::start_time_ = 0;
+KeyAddr OneShot::prev_key_addr_ = UnknownKeyswitchLocation;
+uint8_t OneShot::release_countdown_ = 0;
```

Comment:
Rather than use the boolean value `should_cancel_`, I use this `release_countdown_` value to delay the release of OneShot keys by one cycle so that they will affect the keypress they're intended to.